### PR TITLE
The component's total activity

### PR DESCRIPTION
**Pull request by BotrosHanna-INL**
--------
Pull Request Description
--------
##### What issue does this change request address?

https://github.com/idaholab/HERON/issues/289

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [x] 4. Automated Tests should pass.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [x] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: src/DispatchManager.py [hunk 1]
```diff
@@ -314,9 +325,67 @@ def _do_dispatch(self, meta, all_structure, project_life, interp_years, segs, se
         # build evaluation cash flows
         self._segment_cashflow(meta, s, seg, year, dispatch, multiplicity,
                                project_life, interp_years, all_structure, final_components)
+        cluster_year_list_of_dicts = [] # collect all component activity data for for one cluster for one year here
```
Comment (dylanjm) [inline src/DispatchManager.py]: Could all this added logic be refactored into a separate function? The `_do_dispatch()` function is getting pretty large.

File: src/DispatchManager.py [hunk 2]
```diff
@@ -314,9 +325,67 @@ def _do_dispatch(self, meta, all_structure, project_life, interp_years, segs, se
...
+
+        for step_index in range(len(dispatch._times)):
+          Dict = {}
```
Comment (dylanjm) [inline src/DispatchManager.py]: Can you use a more descriptive variable name here?

File: src/DispatchManager.py [hunk 3]
```diff
@@ -314,9 +325,67 @@ def _do_dispatch(self, meta, all_structure, project_life, interp_years, segs, se
...
+          variable_name_list=[]
+
+          for k in (dispatch._data.keys()):
+            var_name = k.replace("_"+k.split("_")[-1], "");# variable name
+            for ky in (dispatch._resources.keys()):
+              if ky.name == var_name:
+                for index in range (len(dispatch._data[k])):
+                  res_name = (list(((dispatch._resources).get(ky)).keys()))[index]
+                  variable_name = "TotalActivity__"+ "__".join((str(k)).rsplit("_", 1))+ "__"+ res_name
```
Comment (dylanjm) [inline src/DispatchManager.py]: Add whitespace between `+` here

File: src/DispatchManager.py [hunk 4]
```diff
@@ -314,9 +325,67 @@ def _do_dispatch(self, meta, all_structure, project_life, interp_years, segs, se
...
+          tot_activity_series_dict["Year"] = interp_year
+          tot_activity_series_dict["Cluster"] = seg
+          tot_activity_series_dict["Days per year per cluster"] = multiplicity
+        tot_activity_series_dict_list.append(tot_activity_series_dict)
+      tot_activity_over_clusters_years.append(tot_activity_series_dict_list)
+
+    tot_activity_over_clusters_years_result = [element for nestedlist in tot_activity_over_clusters_years for element in nestedlist]
+
+    tot_activity_per_year_list = [] # The total activity per year
+    for y in list_of_years:
```
Comment (dylanjm) [inline src/DispatchManager.py]: Maybe instead of `y` do ` year`

File: src/DispatchManager.py [hunk 5]
```diff
@@ -314,9 +325,67 @@ def _do_dispatch(self, meta, all_structure, project_life, interp_years, segs, se
...
+      keys = ["Year"]+variable_name_list
+      values =[y] + [0] * len(variable_name_list)
+      tot_activity_per_year = dict(zip(keys,values))
+      for d in tot_activity_over_clusters_years_result:
+        if d["Year"] == y:
+          for v in variable_name_list:
+            tot_activity_per_year[v] = tot_activity_per_year[v] + d[v] * d['Days per year per cluster']
```
Comment (dylanjm) [inline src/DispatchManager.py]: I'm curious how much we should hard-code logic relying on clusters when they may go away or are treated differently in the future.

File: templates/template_driver.py [hunk 6]
```diff
@@ -44,6 +43,9 @@
 DEFAULT_STATS_NAMES = ['expectedValue', 'sigma', 'median']
 SWEEP_DEFAULT_STATS_NAMES = ['maximum', 'minimum', 'percentile', 'samples', 'variance']
 
+# prefixes for financial metrics only
+fin_pre = ["sharpe", "sortino", "es", "VaR", "glr"]
```
Comment (dylanjm) [inline templates/template_driver.py]: Make any global variables all-caps and be more descriptive like `FINANCIAL_NAMES` or something.

File: templates/template_driver.py [hunk 7]
```diff
@@ -299,27 +302,37 @@ def _modify_outer_vargroups(self, template, case, components, sources):
...
     econ_metrics = case.get_econ_metrics(nametype='output')
-    has_mult_metrics = len(econ_metrics) > 1
     # loop through all economic metrics (e.g., NPV, IRR) and apply required defaults to each
     default_stats_prefixes = self._get_stats_metrics_prefixes(case, DEFAULT_STATS_NAMES)
     default_stats = [self.namingTemplates['metric_name'].format(stats=sp, econ=em) \
                      for em in econ_metrics for sp in default_stats_prefixes]
-    for stat in default_stats:
+
+    # total activity statistics
+    default_stats_tot_act = [self.namingTemplates['tot_activity'].format(stats=sp, component=component.name, tracker=tracker, resource=resource) for component in components for tracker in component.get_tracking_vars() for resource in component.get_resources() for sp in default_stats_prefixes]
```
Comment (dylanjm) [inline templates/template_driver.py]: I would just break this out into a traditional nested for loop. It's kind of hard to read as a comprehension.

File: templates/template_driver.py [hunk 8]
```diff
@@ -299,27 +302,37 @@ def _modify_outer_vargroups(self, template, case, components, sources):
...
     elif case.get_mode() == 'sweep':
       # loop through all economic metrics (e.g., NPV, IRR) and apply required sweep defaults to each
       sweep_stats_prefixes = self._get_stats_metrics_prefixes(case, DEFAULT_STATS_NAMES+SWEEP_DEFAULT_STATS_NAMES)
       sweep_default = [self.namingTemplates['metric_name'].format(stats=sp, econ=em) \
                       for em in econ_metrics for sp in sweep_stats_prefixes]
-      for sweep_name in sweep_default:
-        if sweep_name not in default_stats:
+
+      # total activity statistics
+      sweep_stats_tot_act = [self.namingTemplates['tot_activity'].format(stats=sp, component=component.name, tracker=tracker, resource=resource) for component in components for tracker in component.get_tracking_vars() for resource in component.get_resources() for sp in sweep_stats_prefixes]
```
Comment (dylanjm) [inline templates/template_driver.py]: Same comment as above.

File: templates/template_driver.py [hunk 9]
```diff
@@ -1217,6 +1230,12 @@ def _modify_inner_econ_metrics(self, template, case):
     """
     # get all economic metrics intended for use in TEAL and reported back
     econ_metrics = case.get_econ_metrics(nametype='output')
+    tot_act_vars = []
+    for component in components:
+      for tracker in component.get_tracking_vars():
```
Comment (dylanjm) [inline templates/template_driver.py:1246]: If you want, you can use `itertools.product` to flatten some of these nested for loops.

File: templates/template_driver.py [hunk 10]
```diff
@@ -1246,27 +1265,35 @@ def _modify_inner_result_statistics(self, template, case):
...
     econ_metrics = case.get_econ_metrics(nametype='output')
-    has_mult_metrics = len(econ_metrics) > 1
+
     # loop through all economic metrics (e.g., NPV, IRR) and apply required defaults to each
     default_stats_prefixes = self._get_stats_metrics_prefixes(case, DEFAULT_STATS_NAMES)
     default_stats = [self.namingTemplates['metric_name'].format(stats=sp, econ=em) \
                      for em in econ_metrics for sp in default_stats_prefixes]
-    for stat in default_stats:
+
+    default_stats_tot_act = [self.namingTemplates['tot_activity'].format(stats=sp, component=component.name, tracker=tracker, resource=resource) for component in components for tracker in component.get_tracking_vars() for resource in component.get_resources() for sp in default_stats_prefixes]
```
Comment (dylanjm) [inline templates/template_driver.py]: Same comment as above

File: templates/template_driver.py [hunk 11]
```diff
@@ -1246,27 +1265,35 @@ def _modify_inner_result_statistics(self, template, case):
...
     # sweep mode has default variable names
     elif case.get_mode() == 'sweep':
       # loop through all economic metrics (e.g., NPV, IRR) and apply required sweep defaults to each
       sweep_stats_prefixes = self._get_stats_metrics_prefixes(case, DEFAULT_STATS_NAMES+SWEEP_DEFAULT_STATS_NAMES)
       sweep_default = [self.namingTemplates['metric_name'].format(stats=sp, econ=em) \
                        for em in econ_metrics for sp in sweep_stats_prefixes]
-      for sweep_name in sweep_default:
-        if sweep_name not in default_stats:
+
+      sweep_stats_tot_act = [self.namingTemplates['tot_activity'].format(stats=sp, component=component.name, tracker=tracker, resource=resource) for component in components for tracker in component.get_tracking_vars() for resource in component.get_resources() for sp in sweep_stats_prefixes]
```
Comment (dylanjm) [inline templates/template_driver.py:1311]: Same comment as above

File: templates/template_driver.py [hunk 12]
```diff
@@ -1537,12 +1586,29 @@ def _updateCommaSeperatedList(self, node, new, position=None, before=None, after
     if organize_economics:
       # all entries with full name and with JUST the economic metric (e.g., NPV, IRR) respectively
       entries = list(x.strip() for x in node.text.split(',')) if node.text is not None else []
-      metric_entries = list(x.split('_')[-1] for x in entries)
+      #metric_entries = list(x.split('_')[-1] for x in entries)
+      metric_entries = []
+      for x in entries:
+        # if not x.startswith("perc_"):
+        #   metric_entry = x.split("_", 1)[1]
```
Comment (dylanjm) [inline templates/template_driver.py]: commented code

**Review by dylanjm — CHANGES_REQUESTED** [review]
Some small changes for you to consider. Let me know your thoughts.

File: src/DispatchManager.py (hunk 1 above)
Comment (BotrosHanna-INL) [reply src/DispatchManager.py]: It is 3 functions now instead of one.

File: src/DispatchManager.py (hunk 2 above)
Comment (BotrosHanna-INL) [reply src/DispatchManager.py]: Done

File: src/DispatchManager.py (hunk 3 above)
Comment (BotrosHanna-INL) [reply src/DispatchManager.py]: Done

File: src/DispatchManager.py (hunk 4 above)
Comment (BotrosHanna-INL) [reply src/DispatchManager.py]: I changed it

File: src/DispatchManager.py (hunk 5 above)
Comment (BotrosHanna-INL) [reply src/DispatchManager.py]: The function I modified was already dependent on clusters so I think for the current time we will rely on them.

**Comment by BotrosHanna-INL** [comment]
Thank you so much, Dylan for taking the time to review this PR. I modified the dispatch_manager and working now on the template_driver changes.

File: templates/template_driver.py (hunk 6 above)
Comment (BotrosHanna-INL) [reply templates/template_driver.py]: Done

File: templates/template_driver.py (hunk 7 above)
Comment (BotrosHanna-INL) [reply templates/template_driver.py]: make sense

File: templates/template_driver.py (hunk 8 above)
Comment (BotrosHanna-INL) [reply templates/template_driver.py]: I replaced it with a traditional nested "for" loop.

File: templates/template_driver.py (hunk 9 above)
Comment (BotrosHanna-INL) [reply templates/template_driver.py:1246]: I did not use the itertools.product before. They seem to be very useful but I will stick now with I am familiar with.

File: templates/template_driver.py (hunk 10 above)
Comment (BotrosHanna-INL) [reply templates/template_driver.py]: I replaced it with a traditional nested "for" loop.

File: templates/template_driver.py (hunk 11 above)
Comment (BotrosHanna-INL) [reply templates/template_driver.py:1311]: I replaced it with a traditional nested "for" loop.

File: templates/template_driver.py (hunk 12 above)
Comment (BotrosHanna-INL) [reply templates/template_driver.py]: removed it

**Comment by BotrosHanna-INL** [comment]
@dylanjm 
I have modified the optimization settings test. It takes less than 300 seconds now.

**Comment by joshua-cogliati-inl** [comment]
> Job [CentOS 8](https://civet.inl.gov/job/1729335/) on [1801220](https://github.com/idaholab/HERON/commit/180122019800c8406dbacb0a1adced4ed8a0e2e1) : invalidated by @dylanjm
> 
> Not sure what happend. Failure in Example Plugins.

```
ImportError: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /home/civet/miniconda3/envs/raven_libraries/lib/python3.10/site-packages/pandas/_libs/window/aggregations.cpython-310-x86_64-linux-gnu.so
```
This looks like a conflict between the library that pandas expects and the library that is on the computer.
Of hand I am not sure why this is happening in the HERON test on Centos 8, but not on the RAVEN test, which is using the same pandas library.
py310h7cbd5c2_1, py310h7cbd5c2_1
https://civet.inl.gov/job/1729335/ versus https://civet.inl.gov/job/1731389/

**Review by dylanjm — APPROVED** [review]
Changes are good to merge.

**Pull request merged**